### PR TITLE
Make container set up parameter server for distributed training

### PR DESCRIPTION
**Pull request by laurenyu**
This change moves the distributed server stuff to be initiated by the container. The user now will specify wanting a parameter server by setting a hyperparameter - this will eventually get abstracted better in the Python SDK.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker_mxnet_container/default_save.py [hunk 1]
```diff
@@ -37,10 +38,11 @@ def save(model_dir, model):
...
         model (mxnet.mod.Module): the module to be saved
     """
-    model.symbol.save(os.path.join(model_dir, SYMBOL_PATH))
-    model.save_params(os.path.join(model_dir, PARAMS_PATH))
-
-    signature = [{'name': data_desc.name, 'shape': [dim for dim in data_desc.shape]}
-                 for data_desc in model.data_shapes]
-    with open(os.path.join(model_dir, SHAPES_PATH), 'w') as f:
-        json.dump(signature, f)
+    if current_host == hosts[0]:
```
Comment (icywang86rui) [inline src/sagemaker_mxnet_container/default_save.py]: Let's add a comment here about the first host is the master.

File: src/sagemaker_mxnet_container/training.py [hunk 2]
```diff
@@ -13,14 +13,69 @@
...
+from retrying import retry
 
 import sagemaker_containers.beta.framework as framework
 
+ROLES = ['worker', 'scheduler', 'server']
+
 logger = logging.getLogger(__name__)
 
 
+def _scheduler_host(hosts):
```
Comment (icywang86rui) [inline src/sagemaker_mxnet_container/training.py]: I think venting this method in the default_save module is fine.

File: src/sagemaker_mxnet_container/training.py [hunk 3]
```diff
@@ -13,14 +13,69 @@
...
+            'PS_VERBOSE': ps_verbose,
+        }
+
+    raise ValueError('Unexpected role: {}'.format(role))
+
+
+def _run_mxnet_process(role, hosts, ps_port, ps_verbose):
+    role_env = os.environ.copy()
+    role_env.update(_env_vars_for_role(role, hosts, ps_port, ps_verbose))
+    return subprocess.Popen("python -c 'import mxnet'", shell=True, env=role_env).pid
```
Comment (icywang86rui) [inline src/sagemaker_mxnet_container/training.py]: We are not really using the pid anywhere. Also does this output anything? do we want to handle failures?

File: src/sagemaker_mxnet_container/training.py [hunk 4]
```diff
@@ -13,14 +13,69 @@
...
+    return socket.gethostbyname(host)
+
+
+def _verify_hosts(hosts):
+    for host in hosts:
+        _host_lookup(host)
+
+
 def train(env):
     logger.info('MXNet training environment: {}'.format(env.to_env_vars()))
```
Comment (icywang86rui) [inline src/sagemaker_mxnet_container/training.py]: run_module prints the training environment as well. do the customers see those twice?

File: src/sagemaker_mxnet_container/training.py [hunk 5]
```diff
@@ -13,14 +13,69 @@
...
+    raise ValueError('Unexpected role: {}'.format(role))
+
+
+def _run_mxnet_process(role, hosts, ps_port, ps_verbose):
+    role_env = os.environ.copy()
+    role_env.update(_env_vars_for_role(role, hosts, ps_port, ps_verbose))
+    return subprocess.Popen("python -c 'import mxnet'", shell=True, env=role_env).pid
+
+
+@retry(stop_max_delay=1000 * 60 * 15, wait_exponential_multiplier=100,
```
Comment (icywang86rui) [inline src/sagemaker_mxnet_container/training.py]: This seems like a awful long time too wait. Do we know that the retry is useful?

File: src/sagemaker_mxnet_container/default_save.py (hunk 1 above)
Comment (laurenyu) [reply src/sagemaker_mxnet_container/default_save.py]: given that I'm moving `scheduler_host()` to this module, using that method will help document the situation

File: src/sagemaker_mxnet_container/training.py (hunk 2 above)
Comment (laurenyu) [reply src/sagemaker_mxnet_container/training.py]: done

File: src/sagemaker_mxnet_container/training.py (hunk 3 above)
Comment (laurenyu) [reply src/sagemaker_mxnet_container/training.py]: nope. removed the return statement.

File: src/sagemaker_mxnet_container/training.py (hunk 5 above)
Comment (laurenyu) [reply src/sagemaker_mxnet_container/training.py]: not sure if we have metrics. these are the same numbers we're using for our current MXNet images.

File: src/sagemaker_mxnet_container/training.py (hunk 4 above)
Comment (laurenyu) [reply src/sagemaker_mxnet_container/training.py]: I've seen it only once when looking at integ tests - it seems as though I don't see the SC output unless it fails at that part

**Pull request merged**